chacra/util.py

```diff
@@ -54,6 +55,80 @@ def repo_paths(repo):
     return paths
 
 
+def get_combined_repos(project, repo_config=None):
+    """
+    Configuration can define specific project repositories to be
+    'combined', this combination of different distro versions are only for
+    Debian-based distros and involves some spelunking in the dictionary
+    to configure them.
+
+    This helper will always return a list because that is the expectation from
+    the configuration.
+    """
+    repo_config = repo_config or getattr(conf, 'repos', {})
+    if not repo_config:
+        return []
+    return repo_config.get(project, {}).get('combined', [])
+
+
+def get_extra_repos(project, ref=None, repo_config=None):
+    """
+    Go through the configuration options for each 'ref' in a project and return
+    the matching ref option for a project, falling to 'all' which signals work
+    for all (but really 'any' in this case) refs.
+
+    If nothing is defined an empty dictionary is returned, so that consumers
+    can treat the return values always as a dictionary
+    """
+    repo_config = repo_config or getattr(conf, 'repos', {})
+    project_ref = ref or 'all'
+    if not repo_config:
+        logging.debug('no repos configuration defined for extra repositories')
+        return {}
+    project_config = repo_config.get(project, {})
+    if not project_config:
+        logging.debug('%s has no configuration for extra repositories', project)
+        return {}
+    extras = project_config.get(project_ref) or project_config.get('all', {})
+    if not extras:
+        logger.warning('%s has no matching repositories for ref: %s', project, project_ref)
+    return extras
+
+
+def get_extra_binaries(project_name, distro, distro_version, distro_versions=None, ref=None):
+    """
+    Try to match a given repository with the distinctive  project/ref/distro
+    information and return a list of associated binaries
+    """
+    binaries = []
+    project = models.Project.query.filter_by(name=project_name).first()
+    if not project:
+        return []
+    repo_query = models.Repo.query.filter_by(project=project)
+
+    if distro_versions:
+        repo_query = repo_query.filter(models.Repo.distro_version.in_(distro_versions))
+    else:
+        repo_query = repo_query.filter_by(distro_version=distro_version)
+
+    if distro is not None:
+        repo_query = repo_query.filter_by(distro=distro)
+    if ref is None:
+        # means that we should just get everything that matches our original
+        # query as a list
+        binaries = []
+        for r in repo_query.all():
+            binaries += [b for b in r.binaries]
+        return binaries
```

Comment:
Should we not return here and instead let this return on line 129 after the logger.info?